com.teamscale.polarion.plugin/src/com/teamscale/polarion/plugin/WorkItemUpdatesServlet.java

```diff
@@ -0,0 +1,769 @@
+   * consider them as items NOT contained in the module. *
+   */
+  private boolean wasMovedToRecycleBin(IWorkItem workItem) {
+    return !module.containsWorkItem(workItem);
+  }
+
+  /**
+   * Not all items in the recycle bin are supposed to be included in the response. It'll depend on
+   * the request lastUpdate revision and endRevision parameters. *
+   */
+  private boolean shouldIncludeItemFromRecybleBin(IWorkItem workItem) {
+    Integer workItemLastRevision = Integer.parseInt(workItem.getLastRevision());
+
+    return (workItemLastRevision > Integer.parseInt(lastUpdate)
+        && workItemLastRevision <= Integer.parseInt(endRevision));
+  }
+
+  /** Create the work item object as DELETED * */
+  private WorkItemForJson buildDeletedWorkItemForJson(IWorkItem workItem) {
+    WorkItemForJson item = new WorkItemForJson(workItem.getId(), Utils.UpdateType.DELETED);
+    // Note: Items in the recycle bin can still undergo changes. For instance, if
+    // any of their field values change, or their links, or even if their module changes id,
+    // it'll generate a new revision and changes will be tracked by Polarion.
+    // Therefore, the following revision will either be the revision when item was deleted or the
+    // the revision when item was lastly changed while in the recycle bin.
+    item.setRevision(workItem.getLastRevision());
+    return item;
+  }
+
+  /** Main method that will process the work item history based on the parameters in the request */
+  private WorkItemForJson processHistory(IWorkItem workItem, IDataService dataService) {
+
+    // Short circuit for performance reasons, don't need to make Polarion fetch history
+    if (Integer.valueOf(workItem.getLastRevision()) <= Integer.valueOf(lastUpdate)) {
+      return null;
+    }
+
+    WorkItemForJson workItemForJson = null;
+
+    IPObjectList<IWorkItem> workItemHistory = dataService.getObjectHistory(workItem);
+    if (workItemHistory != null) {
+      if (workItemHistory.size() == 1 && workItemHistory.get(0) != null) {
+        // No changes in history when size == 1 (the WI remains as created)
+        // We then return only if the item was created within the revision boundaries of the request
+        if (Integer.valueOf(workItemHistory.get(0).getRevision()) <= Integer.valueOf(endRevision))
+          workItemForJson =
+              Utils.castWorkItem(workItemHistory.get(0), includeCustomFields, includeLinkRoles);
+      } else if (workItemHistory.size() > 1) {
+        /**
+         * From Polarion JavaDoc: "The history list is sorted from the oldest (first) to the newest
+         * (last)."
+         * https://almdemo.polarion.com/polarion/sdk/doc/javadoc/com/polarion/platform/persistence/IDataService.html#getObjectHistory(T)
+         * Then, we get the last one from the history as the current revision
+         */
+        int lastUpdateIndex = searchIndexWorkItemHistory(workItemHistory);
+        IDiffManager diffManager = dataService.getDiffManager();
```

Comment:
I would directly check for a valid `lastUpdateIndex` here and return early

Reply:
Done